packages/console/src/hooks/use-position.ts

```diff
@@ -0,0 +1,225 @@
+import { RefObject, useCallback, useEffect, useState } from 'react';
+
+type VerticalAlignment = 'top' | 'bottom';
+
+type HorizontalAlignment = 'start' | 'center' | 'end';
+
+type Offset = {
+  vertical: number;
+  horizontal: number;
+};
+
+type Props = {
+  verticalAlign: VerticalAlignment;
+  horizontalAlign: HorizontalAlignment;
+  offset: Offset;
+  anchorRef: RefObject<Element>;
+  overlayRef: RefObject<Element>;
+  autoAlign?: boolean;
+};
+
+type Position = {
+  top: number;
+  left: number;
+};
+
+// Leave space for box-shadow effect.
+const windowSafePadding = 12;
+
+const getSupportedVerticalAlignments = ({
+  verticalTop,
+  verticalBottom,
+  overlayHeight,
+}: {
+  verticalTop: number;
+  verticalBottom: number;
+  overlayHeight: number;
+}) => {
+  const supportedVerticalAlignments = new Set<VerticalAlignment>(['top', 'bottom']);
+  const minY = windowSafePadding;
+  const maxY = window.innerHeight - windowSafePadding;
+
+  if (verticalTop < minY) {
+    supportedVerticalAlignments.delete('top');
+  }
+
+  if (verticalBottom + overlayHeight > maxY) {
+    supportedVerticalAlignments.delete('bottom');
+  }
+
+  return [...supportedVerticalAlignments];
+};
+
+const getSupportedHorizontalAlignments = ({
+  horizontalStart,
+  horizontalCenter,
+  horizontalEnd,
+  overlayWidth,
+}: {
+  horizontalStart: number;
+  horizontalCenter: number;
+  horizontalEnd: number;
+  overlayWidth: number;
+}) => {
+  const validHorizontalAlignments = new Set<HorizontalAlignment>(['start', 'center', 'end']);
+
+  const minX = windowSafePadding;
+  const maxX = window.innerWidth - windowSafePadding;
+
+  if (horizontalEnd < minX) {
+    validHorizontalAlignments.delete('end');
+  }
+
+  if (horizontalCenter - overlayWidth / 2 < minX) {
+    validHorizontalAlignments.delete('center');
+  }
+
+  if (horizontalCenter + overlayWidth / 2 > maxX) {
+    validHorizontalAlignments.delete('center');
+  }
+
+  if (horizontalStart + overlayWidth > maxX) {
+    validHorizontalAlignments.delete('start');
+  }
+
+  return [...validHorizontalAlignments];
```

Comment:
These two functions are a little bit strange. Why do we need to get supported horizontal and vertical alighments in the first place??
And `delete` function is a mutable function, I believe, which is also not recommended to use...
I'll see if others have any ideas on this one

Reply:
This function is used to filter the unsupported alignment when we use `autoAlign`.

Reply:
Any good alternatives to this `delete` logic?

Reply:
Is autoAlign a required feature? When should we use `autoAlign`?

Reply:
This is for compatibility with `Dropdown`.

Reply:
updated.